File: src/integration/wirespec/src/jvmMain/java/community/flock/wirespec/java/serde/DefaultPathSerialization.java
package community.flock.wirespec.java.serde;

import community.flock.wirespec.java.Wirespec;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Type;

import static community.flock.wirespec.java.serde.DefaultSerialization.*;

public interface DefaultPathSerialization extends Wirespec.PathSerialization {

    @Override
    default <T> String serializePath(T t, Type type) {
        if(t instanceof Wirespec.Refined<?> refined){
            return refined.value().toString();
        }
        if(t instanceof Wirespec.Enum refined){
            return refined.label();
        }
        return t.toString();
    }

    @Override
    @SuppressWarnings("unchecked")
    default  <T> T deserializePath(String raw, Type type) {
        final Class<?> rawType = getRawType(type);
        if(isWirespecRefined(rawType)) {
            try {
                Constructor<?> constructor = rawType.getDeclaredConstructors()[0];
                Class<?> paramType = constructor.getParameterTypes()[0];
                Object value = PRIMITIVE_TYPES_CONVERSION.get(paramType).apply(raw);
                return (T) constructor.newInstance(value);
            } catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
                throw new RuntimeException(e);
            }
        }
        if(isWirespecEnum(rawType)) {
            return (T) findEnumByLabel(rawType, raw);
        }
        return (T) PRIMITIVE_TYPES_CONVERSION.get(rawType).apply(raw);
    }
}

File: src/integration/wirespec/src/jvmMain/kotlin/community/flock/wirespec/kotlin/serde/DefaultPathSerialization.kt
package community.flock.wirespec.kotlin.serde

import community.flock.wirespec.kotlin.Wirespec
import community.flock.wirespec.kotlin.serde.DefaultSerialization.classifierAsKClass
import community.flock.wirespec.kotlin.serde.DefaultSerialization.findEnumByLabel
import community.flock.wirespec.kotlin.serde.DefaultSerialization.isWirespecEnum
import community.flock.wirespec.kotlin.serde.DefaultSerialization.isWirespecRefined
import community.flock.wirespec.kotlin.serde.DefaultSerialization.primitiveTypesConversion
import kotlin.reflect.KClass
import kotlin.reflect.KType
import kotlin.reflect.full.primaryConstructor

public class DefaultPathSerialization : Wirespec.PathSerialization {
    override fun <T : Any> serializePath(t: T, type: KType): String = when (t) {
        is Wirespec.Refined<*> -> t.value.toString()
        is Wirespec.Enum -> t.label
        else -> t.toString()
    }

    @Suppress("UNCHECKED_CAST")
    override fun <T : Any> deserializePath(raw: String, type: KType): T = when {
        type.isWirespecRefined() -> {
            val kClass = type.classifierAsKClass()
            val constructor = kClass.primaryConstructor
                ?: error("No primary constructor found for refined type: ${kClass.simpleName}")
            val paramType = constructor.parameters.first().type.classifier as KClass<*>
            val convertedValue = primitiveTypesConversion[paramType]?.invoke(raw)
                ?: error("Unsupported refined value type: ${paramType.simpleName}")
            constructor.call(convertedValue)
        }
        type.isWirespecEnum() -> findEnumByLabel(type.classifierAsKClass(), raw)
        else -> primitiveTypesConversion[type.classifierAsKClass()]?.invoke(raw)
    } as T
}
